src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -0,0 +1,327 @@
+import net.minecraft.entity.monster.EntitySlime;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.EntityEquipmentSlot;
+import net.minecraft.item.*;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import static com.minecolonies.entity.ai.util.AIState.GUARD_RESTOCK;
+import static com.minecolonies.entity.ai.util.AIState.IDLE;
+import static com.minecolonies.entity.ai.util.AIState.START_WORKING;
+
+/**
+ * Abstract class which contains all the guard basics let it be range, melee or magic.
+ */
+public abstract class AbstractEntityAIGuard extends AbstractEntityAISkill<JobGuard>
+{
+    /**
+     * The start search distance of the guard to track/attack entities may get more depending on the level.
+     */
+    private static final double MAX_ATTACK_DISTANCE = 20.0D;
+
+    /**
+     * Distance the guard starts searching.
+     */
+    private static final int START_SEARCH_DISTANCE = 5;
+
+    /**
+     * Basic delay after operations.
+     */
+    private static final int BASE_DELAY = 1;
+
+    /**
+     * Max amount the guard can shoot arrows before restocking.
+     */
+    private static final int MAX_ATTACKS = 50;
+
+    /**
+     * Y range in which the guard detects other entities.
+     */
+    private static final double HEIGHT_DETECTION_RANGE = 10D;
+
+    /**
+     * Path that close to the patrol target.
+     */
+    private static final int PATH_CLOSE                = 3;
+
+    /**
+     * The distance the guard is searching entities in currently.
+     */
+    private int currentSearchDistance = START_SEARCH_DISTANCE;
+
+    /**
+     * The current target.
+     */
+    protected EntityLivingBase targetEntity;
+
+    /**
+     * Current goTo task.
+     */
+    private BlockPos currentPatrolTarget;
+
+    /**
+     * Containing all close entities.
+     */
+    private List<Entity> entityList;
+
+    /**
+     * Amount of arrows already shot or sword hits dealt.
+     */
+    protected int attacksExecuted = 0;
+
+    /**
+     * Sets up some important skeleton stuff for every ai.
+     *
+     * @param job the job class
+     */
+    protected AbstractEntityAIGuard(@NotNull final JobGuard job)
+    {
+        super(job);
+        super.registerTargets(
+                new AITarget(IDLE, () -> START_WORKING),
+                new AITarget(START_WORKING, () -> GUARD_RESTOCK)
+        );
+    }
+
+
+    /**
+     * Can be overridden in implementations.
+     * <p>
+     * Here the AI can check if the armour have to be re rendered and do it.
+     */
+    @Override
+    protected void updateRenderMetaData()
+    {
+        updateArmor();
+    }
+
+    /**
+     * Goes back to the building and tries to take armour from it when he hasn't in his inventory.
+     * @return the next state to go to.
+     */
+    protected AIState goToBuilding()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "goToBuilding" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)